Add Multivariate Distance Covariance metric and the corresponding test of independence

Palash123-4 opened the pull request:
As we discussed, I have added the code corresponding to the random projection based multivariate distance covariance, as discussed in https://www.frontiersin.org/articles/10.3389/fams.2021.779841/full.
I spent some time regarding how to use the gamma function in the best possible way so that the correct p_value can be calculated. (One need to install the "mpmath" python module for gamma_cdf function, for details:  https://mpmath.org/doc/current/functions/expintegrals.html)
Please take a look and let me know whether it is working fine or not.

Palash123-4 committed:
Add files via upload

Palash123-4 committed:
Add the multivariate distance covariance

Palash123-4 committed:
Update _dcor.py

vnmabus commented on dcor/independence.py:
```diff
@@ -12,6 +12,15 @@
+## Additional modules for Multivariate dcov-based test of independence------------
+import math
+from ._dcor import u_distance_covariance_sqr, dist_sum, gamma_ratio, rndm_projection  
+from mpmath import*
```
Do not do a wildcard import outside of a console session. It is bad practice (obscures the code and may cause difficult-to-debug errors).
`mpmath` is not a dependence of this package, and should not be necessary.

Palash123-4 replied:
I have commented this part

vnmabus commented on dcor/independence.py:
```diff
@@ -347,3 +356,105 @@ def distance_correlation_t_test(
+
+mp.dps = 25; mp.pretty = True
+def gamma_cdf(x, shape,  scale):
+    return gammainc(shape, a = 0, b = float(x/scale))/np.exp(gammaln(shape))
```
You can use [`scipy.special.gammainc`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.gammainc.html), so that `mpmath` is not required.

Palash123-4 replied:
I have changed the code as you suggested, but based on my experimental outcomes mpmath helps to calculate accurate integration compare to scipy.stats one

vnmabus reviewed (changes requested):
Thank you for your effort making this PR! I did not check the math yet (I assume you implemented it well), but I have several code suggestions.

Palash123-4 committed:
Merge branch 'vnmabus:develop' into develop

Commit:
update random projection-
based distance covariance

Commit:
update random projection based test of independence

Commit:
add dist_sum function

Commit:
Merge branch 'develop' of https://github.com/Palash123-4/dcor into develop

Commit:
add citation of random projections based distance covariance

Palash123-4 commented on dcor/_dcor.py:
```diff
@@ -1169,3 +1174,128 @@ def distance_correlation_af_inv(
+# `np.bool` was a deprecated alias for the builtin `bool`. To avoid this error in existing code, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
+# The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
+#     https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
+np.bool = np.bool_
```
I have removed it

Palash123-4 commented on dcor/independence.py:
```diff
@@ -347,3 +356,105 @@ def distance_correlation_t_test(
+    omega3_n = 0
+
+    # for i in tqdm(range(n_projs)):
+    for i in range(n_projs):
```
Thanks for your suggestion, vectorization works

Palash123-4 commented on dcor/_dcor.py:
```diff
@@ -1169,3 +1174,128 @@ def distance_correlation_af_inv(
+        >>> import dcor
+        >>> from scipy.stats import multivariate_normal
+        >>> mean_vector = [2, 3, 5, 3, 2, 1]
+        >>> matrixSize = 6
```
I have changed it

Palash123-4 commented on dcor/_dcor.py:
```diff
@@ -1169,3 +1174,128 @@ def distance_correlation_af_inv(
+    X_std = np.random.standard_normal(dim_X)
+    X_norm = np.sum([i**2 for i in X_std])
+    U_sphere = np.array(X_std / np.sqrt(X_norm))
+    if dim_X > 1:
```
The idea of considering == 1 is, one may interest to compute the distance correlation between a univariate and a multivariate data

Palash123-4 replied:
I have taken a look at the thesis of the main author of random projection paper:
https://repository.gatech.edu/server/api/core/bitstreams/27115b8a-5875-445a-a972-b1686799ecb5/content

Based on my understanding from that thesis, the test of independence based on both "dcov(X, Y)(naive)" and "dcov(X, Y)(random projection)" could work on the general data (i.e., dimension of X and Y >=1). The only difference is "dcov(X,Y)(random projection)" based test is faster than "dcov(X,Y)(naive)" test of independence. Now, the question is, can we calculate random projection "dcov (using random projection)" and implement the naive test? Answer is no because one is based on biased dcov estimator and another one is based on unbiased estimator. Moreover, as unbiased dcov does not imply unbiased dcorr (ratio of two unbiased estimators does not guarantee an unbiased estimator of ratio), we cannot even do the multivariate t-test (existing multivariate tests). We have to use a separate test where large sample distribution of the test statistic follows Gamma Distribution (as proposed by the random projection paper).

Palash123-4 commented on dcor/_dcor.py:
```diff
@@ -1169,3 +1174,128 @@ def distance_correlation_af_inv(
+    sqrt_pi_value = math.sqrt(math.pi)
+    # C_p = (math.sqrt(pi_value)*gamma((p+1)/2))/ gamma(p/2)
+    # C_q = (math.sqrt(pi_value)*gamma((q+1)/2))/ gamma(q/2)
+    C_p = sqrt_pi_value*gamma_ratio(p)
```
I have updated with addition of all the spaces

Palash123-4 commented on dcor/independence.py:
```diff
@@ -347,3 +356,105 @@ def distance_correlation_t_test(
+url: https://www.frontiersin.org/articles/10.3389/fams.2021.779841/full
+'''
+
+mp.dps = 25; mp.pretty = True
```
I have commented it

Commit:
update with spaces

Commit:
update with spaces

Commit:
update with spaces

Commit:
insert .

Commit:
insert .

Commit:
insert .

Commit:
insert .

Commit:
update module

Commit:
update module

Palash123-4 commented on dcor/_dcor.py:
```diff
@@ -1169,3 +1174,128 @@ def distance_correlation_af_inv(
+np.bool = np.bool_
+
+
+def gamma_ratio(p): return np.exp(gammaln((p+1)/2) - gammaln(p/2))  # For Calculating C_p and C_q
```
I have added the type-annotation

Palash123-4 replied:
I have made the changes

Palash123-4 commented on dcor/_dcor.py:
```diff
@@ -1169,3 +1174,128 @@ def distance_correlation_af_inv(
+    return X_new
+
+
+def u_dist_cov_sqr_mv(X, Y, p, q, n_projs= 500, fast_method='mergesort', func=u_distance_covariance_sqr):
```
I have made the changes.

How is the default for n_projs selected?
- The larger the number of projections, the better is the estimated value. The paper https://www.frontiersin.org/articles/10.3389/fams.2021.779841/full mentioned that a reasonable consideration of number of projections is (N/log(N)), where N is the number of samples. In my implementation I have considered 500, where I saw in some of their experiments they have considered n_projs = 50

Palash123-4 commented:
I found some issue regarding the cyclic recalling of rowwise module in _dcor.py. Maybe you can help me to solve it.

Palash123-4 commented on dcor/_dcor.py:
```diff
@@ -1169,3 +1174,128 @@ def distance_correlation_af_inv(
+    '''
+
+    # X_std = multivariate_normal.rvs( np.zeros(dim_X), np.identity(dim_X), size = 1)
+    X_std = np.random.standard_normal(dim_X)
```
We need random projections, that means for each projection we need fresh randomly generated values from unite sphere of dimension p, for a fixed data, u_dist_cov_sqr_mv always provide same values if we consider a higher number of projections

Palash123-4 replied:
Briefly, the idea is:
1.  I have to draw u, v from unite sphere  (where I consider first standard normal and normalize it).
2. Then I have to evaluate the projected value (u^TX_i, v^TY_i ) for i = 1,2,...., n and finally calculate the "dcov".

I have to repeat the steps 1 and 2 for K (number of projections) times, Now if I consider a fixed random state at step 1, Then at step 2, I will always get the same "dcov" value, which implies for K= 1 and K = 50 I always get me the same value of final "dcov" value (as it is equal to the mean of all K "dcov"s which I evaluated at step 2)  which might results to a biased estimator. As with increase the sample size, sample mean converges to a fixed value (or within a small interval), I am certain that the reproducibility of the final results will not affect that much until we have considered a very small value of K.

This is my thinking. Please correct me if I am wrong in any case

Commit:
update dimension evaluation

Commit:
update dimension evaluation

Palash123-4 commented on dcor/_dcor.py:
```diff
@@ -1169,3 +1174,128 @@ def distance_correlation_af_inv(
+
+
+@njit(fastmath=True, parallel=True, cache=True)
+def dist_sum(X):  # It is nothing but the parallelized version of pdist function-----------------------
```
I have added dist_sum in the distances module

Palash123-4 commented on dcor/_dcor.py:
```diff
@@ -1169,3 +1174,128 @@ def distance_correlation_af_inv(
+@njit(fastmath=True, parallel=True, cache=True)
+def dist_sum(X):  # It is nothing but the parallelized version of pdist function-----------------------
+    '''
+    Parameters
```
I have replaced ' ' ' by " " "

vnmabus replied:
Google style docstrings, as opposed to NumPy style docstrings, use a colon to specify sections such as `Args:` or `Returns:`, instead of placing hyphens in the line below. See https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html for examples.

Palash123-4 commented on dcor/_dcor.py:
```diff
@@ -1169,3 +1174,128 @@ def distance_correlation_af_inv(
+
+    # X_std = multivariate_normal.rvs( np.zeros(dim_X), np.identity(dim_X), size = 1)
+    X_std = np.random.standard_normal(dim_X)
+    X_norm = np.sum([i**2 for i in X_std])
```
I made the changes in the present code

Palash123-4 commented:
Re: Maybe, but does the general code not work in that case?

I have taken a look at the thesis of the main author of random projection paper:
https://repository.gatech.edu/server/api/core/bitstreams/27115b8a-5875-445a-a972-b1686799ecb5/content

Based on my understanding from that thesis, the test of independence based on both "dcov(X, Y)(naive)" and "dcov(X, Y)(random projection)" could work on the general data (i.e., dimension of X and Y >=1). The only difference is "dcov(X,Y)(random projection)" based test is faster than "dcov(X,Y)(naive)" test of independence. Now, the question is, can we calculate random projection "dcov (using random projection)" and implement the naive test? Answer is no because one is based on biased dcov estimator and another one is based on unbiased estimator. Moreover, as unbiased dcov does not imply unbiased dcorr (ratio of two unbiased estimators does not guarantee an unbiased estimator of ratio), we cannot even do the multivariate t-test (existing multivariate tests). We have to use a separate test where large sample distribution of the test statistic follows Gamma Distribution (as proposed by the random projection paper).

Palash123-4 commented on dcor/_dcor.py:
```diff
@@ -1169,3 +1175,131 @@ def distance_correlation_af_inv(
+
+References
+----------
+.. bibliography:: ../refs.bib
```
I would really appreciate it if you kindly guide me how to do that, I did search this names on the internet, but I am not quite certain how to do it. It would be really nice if you guide me step-by-step process.

vnmabus replied:
Just use `:footcite:` instead of `:cite:` inside the text to place the footnote, and `.. footbibliography::` instead of `.. biblioggraphy::`. You do not need to specify the .bib file, as it is done in the global config.

You can use this docstring as an example:
https://github.com/GAA-UAM/scikit-fda/blob/3eb8ace81220011559df83be0ffb9154a11fb001/skfda/exploratory/stats/_fisher_rao.py#L66-L99

Palash123-4 replied:
I implemented the exact format that you have mentioned

Palash123-4 committed:
Add files via upload

Palash123-4 commented on dcor/_dcor.py:
```diff
@@ -1169,3 +1175,131 @@ def distance_correlation_af_inv(
+    return X_new
+
+
+def u_dist_cov_sqr_mv(X, Y, n_projs = 500, method ='mergesort'):
```
I made the changes

Palash123-4 commented on dcor/_dcor.py:
```diff
@@ -1169,3 +1175,131 @@ def distance_correlation_af_inv(
+
+
+"""
+A Statistically and Numerically Efficient Independence Test Based on
```
I completed the task

vnmabus commented:
Note that the last commit shows a lot of whitespace changes. Please, make sure that your editor is not changing the Unix newlines (LF, or "\n") to Windows ones (CRLF, or "\r\n").

Palash123-4 commented:
> Note that the last commit shows a lot of whitespace changes. Please, make sure that your editor is not changing the Unix newlines (LF, or "\n") to Windows ones (CRLF, or "\r\n").

I use actually "spyder", not sure why it did happen. But I can use "autopipe" for better indentation if it's necessary. 
Can you please tell me why am I getting this compilation error?

Palash123-4 committed:
Add files via upload

Removed White space

Palash123-4 commented:
> Note that the last commit shows a lot of whitespace changes. Please, make sure that your editor is not changing the Unix newlines (LF, or "\n") to Windows ones (CRLF, or "\r\n").

I have removed the white space.

Palash123-4 committed:
Merge branch 'vnmabus:develop' into develop

Palash123-4 committed:
Fix circular import in _rowwise module initialization

- Changed default argument from _dcor.DistanceCovarianceMethod.AUTO to None to avoid circular import
- Created _initialize_rowwise_functions() to delay module-level function assignments until all imports complete
- Call initialization function after all modules are loaded in __init__.py
- Added 'as' clause for partial_distance_correlation import for consistency

Palash123-4 committed:
Merge branch 'develop' into develop

Palash123-4 committed:
Fix ReadTheDocs build by using importlib.metadata instead of pkg_resources

- Replace deprecated pkg_resources with importlib.metadata (Python 3.8+)
- This fixes ModuleNotFoundError in Sphinx build environment on ReadTheDocs
- importlib.metadata is the modern, standard library replacement

Palash123-4 committed:
Merge branch 'develop' of https://github.com/Palash123-4/dcor into develop

Palash123-4 committed:
Fix BibTeX syntax error in refs.bib

- Added missing comma after url field in dcov_random_projection entry
- BibTeX requires commas between all fields except the last one
- This fixes ReadTheDocs build error: pybtex.scanner.TokenRequired: syntax error in line 11
